state-chain/chains/src/btc/smart_contract_encoding.rs

```diff
@@ -0,0 +1,95 @@
+use crate::ForeignChainAddress;
+use cf_primitives::{Asset, AssetAmount};
+use codec::{Decode, Encode};
+
+use super::{BitcoinOp, BitcoinScript};
+
+// The maximum length of data that can be encoded in a nulldata utxo
+const MAX_NULLDATA_UTXO_LENGTH: usize = 80;
+
+#[derive(Encode, Decode, Clone)]
+pub struct UtxoEncodedData {
+	pub output_asset: Asset,
+	pub output_address: ForeignChainAddress,
+	pub parameters: SharedCfParameters,
+}
+
+// The encoding of these parameters is the same across chains
+#[derive(Encode, Decode, Clone)]
+pub struct SharedCfParameters {
+	// FoK fields (refund address is stored externally):
+	pub retry_duration: u16,
+	pub min_output_amount: AssetAmount,
+	// DCA fields:
+	pub number_of_chunks: u16,
+	pub chunk_interval: u16,
+	// Boost fields:
+	pub boost_fee: u8,
+}
+
+#[allow(dead_code)]
+pub fn encode_data_in_nulldata_utxo(data: &[u8]) -> Option<BitcoinScript> {
+	if data.len() > MAX_NULLDATA_UTXO_LENGTH {
+		return None;
+	}
```

Comment:
This shouldn't be possible. We should debug_assert this and/or fail loudly instead of silently dropping it.

Reply:
The funciton accepts `&[u8]` from whoever is calling it, so it is very much possible. Also, how is possible to "silently drop" the result of encoding? Note that return `None` which should be handled by the caller to get the result. It is the caller's job to assert (with debug_asert or otherwise) that encoding worked.

Of course we can assert when passing the result of `params.encode()` that the result of encoding `UtxoEncodedData` with Scale always fits within 80 bytes. I was a little hesitant to do this since, it is easy to add another field or a variant to an existing type that would result in exceeding the max length. (We have tests for this of course, but it is hard to test exhaustively). Seems better to return Error to the consumer of this for a particular set of parameters than to crash the broker? Anyway, I assert internally that `params.encode()` fits, but I think a lot of this fine-tuning of the interfaces would make more sense when we add components that use them.